File: Python_Lang/hospital.py
class Paciente:
    def __init__(self, id, nome, idade, endereco, telefone):
        self.id = id
        self.nome = nome
        self.idade = idade
        self.endereco = endereco
        self.telefone = telefone
        self.prox = None

class Medico:
    def __init__(self, id, nome, especialidade):
        self.id = id
        self.nome = nome
        self.especialidade = especialidade
        self.prox = None

class Consulta:
    def __init__(self, idPaciente, idMedico, gravidade, data):
        self.idPaciente = idPaciente
        self.idMedico = idMedico
        self.gravidade = gravidade
        self.data = data
        self.prox = None

class FilaPrioridade:
    def __init__(self):
        self.inicio = None

    def agendar_consulta(self, idPaciente, idMedico, gravidade, data):
        nova = Consulta(idPaciente, idMedico, gravidade, data)
        if not self.inicio or self.inicio.gravidade < gravidade:
            nova.prox = self.inicio
            self.inicio = nova
        else:
            atual = self.inicio
            while atual.prox and atual.prox.gravidade >= gravidade:
                atual = atual.prox
            nova.prox = atual.prox
            atual.prox = nova
        return nova

class Hospital:
    def __init__(self):
        self.pacientes = None
        self.medicos = None
        self.fila = FilaPrioridade()

    def cadastrar_paciente(self, id, nome, idade, endereco, telefone):
        novo = Paciente(id, nome, idade, endereco, telefone)
        novo.prox = self.pacientes
        self.pacientes = novo
        return novo

    def cadastrar_medico(self, id, nome, especialidade):
        novo = Medico(id, nome, especialidade)
        novo.prox = self.medicos
        self.medicos = novo
        return novo

    def buscar_paciente_por_id(self, id):
        atual = self.pacientes
        while atual:
            if atual.id == id:
                return atual
            atual = atual.prox
        return None

    def buscar_medico_por_id(self, id):
        atual = self.medicos
        while atual:
            if atual.id == id:
                return atual
            atual = atual.prox
        return None

    def listar_pacientes(self):
        print("Lista de Pacientes:")
        atual = self.pacientes
        while atual:
            print(f"ID: {atual.id}, Nome: {atual.nome}, Idade: {atual.idade}, Endereço: {atual.endereco}, Telefone: {atual.telefone}")
            atual = atual.prox

    def listar_medicos(self):
        print("Lista de Médicos:")
        atual = self.medicos
        while atual:
            print(f"ID: {atual.id}, Nome: {atual.nome}, Especialidade: {atual.especialidade}")
            atual = atual.prox

    def listar_consultas(self):
        print("Lista de Consultas:")
        atual = self.fila.inicio
        while atual:
            print(f"Paciente ID: {atual.idPaciente}, Médico ID: {atual.idMedico}, Gravidade: {atual.gravidade}, Data: {atual.data}")
            atual = atual.prox

    def excluir_paciente(self, id):
        atual = self.pacientes
        anterior = None
        while atual:
            if atual.id == id:
                if anterior:
                    anterior.prox = atual.prox
                else:
                    self.pacientes = atual.prox
                print(f"Paciente ID {id} excluído com sucesso.")
                return
            anterior = atual
            atual = atual.prox
        print(f"Paciente ID {id} não encontrado.")

    def excluir_medico(self, id):
        atual = self.medicos
        anterior = None
        while atual:
            if atual.id == id:
                if anterior:
                    anterior.prox = atual.prox
                else:
                    self.medicos = atual.prox
                print(f"Médico ID {id} excluído com sucesso.")
                return
            anterior = atual
            atual = atual.prox
        print(f"Médico ID {id} não encontrado.")

    def excluir_consulta(self, idPaciente, idMedico):
        atual = self.fila.inicio
        anterior = None
        while atual:
            if atual.idPaciente == idPaciente and atual.idMedico == idMedico:
                if anterior:
                    anterior.prox = atual.prox
                else:
                    self.fila.inicio = atual.prox
                print(f"Consulta do Paciente ID {idPaciente} com Médico ID {idMedico} excluída com sucesso.")
                return
            anterior = atual
            atual = atual.prox
        print(f"Consulta não encontrada para o Paciente ID {idPaciente} e Médico ID {idMedico}.")

    def gerar_relatorio(self):
        with open('relatorio.txt', 'w') as fp:
            fp.write("Relatório de Pacientes:\n")
            paciente = self.pacientes
            while paciente:
                fp.write(f"ID: {paciente.id}, Nome: {paciente.nome}, Idade: {paciente.idade}, Endereço: {paciente.endereco}, Telefone: {paciente.telefone}\n")
                paciente = paciente.prox

            fp.write("\nRelatório de Médicos:\n")
            medico = self.medicos
            while medico:
                fp.write(f"ID: {medico.id}, Nome: {medico.nome}, Especialidade: {medico.especialidade}\n")
                medico = medico.prox

            fp.write("\nRelatório de Consultas:\n")
            consulta = self.fila.inicio
            while consulta:
                fp.write(f"Paciente ID: {consulta.idPaciente}, Médico ID: {consulta.idMedico}, Gravidade: {consulta.gravidade}, Data: {consulta.data}\n")
                consulta = consulta.prox

        print("Relatório gerado com sucesso em 'relatorio.txt'.")

# Função principal para o menu
def main():
    hospital = Hospital()

    while True:
        print("\nSistema de Gerenciamento Hospitalar")
        print("1. Cadastrar Paciente")
        print("2. Cadastrar Médico")
        print("3. Agendar Consulta")
        print("4. Listar Pacientes")
        print("5. Listar Médicos")
        print("6. Listar Consultas")
        print("7. Excluir Paciente")
        print("8. Excluir Médico")
        print("9. Excluir Consulta")
        print("10. Gerar Relatório")
        print("11. Sair")
        opcao = int(input("Escolha uma opção: "))

        if opcao == 1:
            id = int(input("ID do Paciente: "))
            nome = input("Nome: ")
            idade = int(input("Idade: "))
            endereco = input("Endereço: ")
            telefone = input("Telefone: ")
            hospital.cadastrar_paciente(id, nome, idade, endereco, telefone)
        elif opcao == 2:
            id = int(input("ID do Médico: "))
            nome = input("Nome: ")
            especialidade = input("Especialidade: ")
            hospital.cadastrar_medico(id, nome, especialidade)
        elif opcao == 3:
            idPaciente = int(input("ID do Paciente: "))
            idMedico = int(input("ID do Médico: "))
            gravidade = int(input("Gravidade (1 a 5): "))
            data = input("Data (dd/mm/aaaa): ")
            hospital.fila.agendar_consulta(idPaciente, idMedico, gravidade, data)
        elif opcao == 4:
            hospital.listar_pacientes()
        elif opcao == 5:
            hospital.listar_medicos()
        elif opcao == 6:
            hospital.listar_consultas()
        elif opcao == 7:
            id = int(input("ID do Paciente: "))
            hospital.excluir_paciente(id)
        elif opcao == 8:
            id = int(input("ID do Médico: "))
            hospital.excluir_medico(id)
        elif opcao == 9:
            idPaciente = int(input("ID do Paciente: "))
            idMedico = int(input("ID do Médico: "))
            hospital.excluir_consulta(idPaciente, idMedico)
        elif opcao == 10:
            hospital.gerar_relatorio()
        elif opcao == 11:
            print("Saindo...")
            break
        else:
            print("Opção inválida. Tente novamente.")

if __name__ == "__main__":
    main()

File: C_Lang/hospital.c
#include "hospital.h"

Paciente* cadastrarPaciente(Paciente **lista, int id, const char *nome, int idade, const char *endereco, const char *telefone) {
    Paciente *novo = (Paciente *)malloc(sizeof(Paciente));
    if (!novo) {
        printf("Erro ao alocar memória para paciente.\n");
        return NULL;
    }
    novo->id = id;
    strncpy(novo->nome, nome, MAX_NAME);
    novo->idade = idade;
    strncpy(novo->endereco, endereco, MAX_NAME);
    strncpy(novo->telefone, telefone, MAX_NAME);
    novo->prox = *lista;
    *lista = novo;
    return novo;
}

Medico* cadastrarMedico(Medico **lista, int id, const char *nome, const char *especialidade) {
    Medico *novo = (Medico *)malloc(sizeof(Medico));
    if (!novo) {
        printf("Erro ao alocar memória para médico.\n");
        return NULL;
    }
    novo->id = id;
    strncpy(novo->nome, nome, MAX_NAME);
    strncpy(novo->especialidade, especialidade, MAX_NAME);
    novo->prox = *lista;
    *lista = novo;
    return novo;
}

Consulta* agendarConsulta(FilaPrioridade *fila, int idPaciente, int idMedico, int gravidade, const char *data) {
    Consulta *nova = (Consulta *)malloc(sizeof(Consulta));
    if (!nova) {
        printf("Erro ao alocar memória para consulta.\n");
        return NULL;
    }
    nova->idPaciente = idPaciente;
    nova->idMedico = idMedico;
    nova->gravidade = gravidade;
    strncpy(nova->data, data, MAX_NAME);
    nova->prox = NULL;

    if (!fila->inicio || fila->inicio->gravidade < gravidade) {
        nova->prox = fila->inicio;
        fila->inicio = nova;
    } else {
        Consulta *atual = fila->inicio;
        while (atual->prox && atual->prox->gravidade >= gravidade) {
            atual = atual->prox;
        }
        nova->prox = atual->prox;
        atual->prox = nova;
    }

    return nova;
}

Paciente* buscarPacientePorID(Paciente *lista, int id) {
    while (lista) {
        if (lista->id == id) {
            return lista;
        }
        lista = lista->prox;
    }
    return NULL;
}

Medico* buscarMedicoPorID(Medico *lista, int id) {
    while (lista) {
        if (lista->id == id) {
            return lista;
        }
        lista = lista->prox;
    }
    return NULL;
}

void listarPacientes(Paciente *lista) {
    printf("Lista de Pacientes:\n");
    while (lista) {
        printf("ID: %d, Nome: %s, Idade: %d, Endereço: %s, Telefone: %s\n", lista->id, lista->nome, lista->idade, lista->endereco, lista->telefone);
        lista = lista->prox;
    }
}

void listarMedicos(Medico *lista) {
    printf("Lista de Médicos:\n");
    while (lista) {
        printf("ID: %d, Nome: %s, Especialidade: %s\n", lista->id, lista->nome, lista->especialidade);
        lista = lista->prox;
    }
}

void listarConsultas(FilaPrioridade *fila) {
    Consulta *atual = fila->inicio;
    printf("Lista de Consultas:\n");
    while (atual) {
        printf("Paciente ID: %d, Médico ID: %d, Gravidade: %d, Data: %s\n", atual->idPaciente, atual->idMedico, atual->gravidade, atual->data);
        atual = atual->prox;
    }
}

void excluirPaciente(Paciente **lista, int id) {
    Paciente *atual = *lista;
    Paciente *anterior = NULL;

    while (atual) {
        if (atual->id == id) {
            if (anterior) {
                anterior->prox = atual->prox;
            } else {
                *lista = atual->prox;
            }
            free(atual);
            printf("Paciente ID %d excluído com sucesso.\n", id);
            return;
        }
        anterior = atual;
        atual = atual->prox;
    }
    printf("Paciente ID %d não encontrado.\n", id);
}

void excluirMedico(Medico **lista, int id) {
    Medico *atual = *lista;
    Medico *anterior = NULL;

    while (atual) {
        if (atual->id == id) {
            if (anterior) {
                anterior->prox = atual->prox;
            } else {
                *lista = atual->prox;
            }
            free(atual);
            printf("Médico ID %d excluído com sucesso.\n", id);
            return;
        }
        anterior = atual;
        atual = atual->prox;
    }
    printf("Médico ID %d não encontrado.\n", id);
}

void excluirConsulta(FilaPrioridade *fila, int idPaciente, int idMedico) {
    Consulta *atual = fila->inicio;
    Consulta *anterior = NULL;

    while (atual) {
        if (atual->idPaciente == idPaciente && atual->idMedico == idMedico) {
            if (anterior) {
                anterior->prox = atual->prox;
            } else {
                fila->inicio = atual->prox;
            }
            free(atual);
            printf("Consulta do Paciente ID %d com Médico ID %d excluída com sucesso.\n", idPaciente, idMedico);
            return;
        }
        anterior = atual;
        atual = atual->prox;
    }
    printf("Consulta não encontrada para o Paciente ID %d e Médico ID %d.\n", idPaciente, idMedico);
}

void gerarRelatorio(Paciente *listaPacientes, Medico *listaMedicos, FilaPrioridade *fila) {
    FILE *fp = fopen("relatorio.txt", "w");
    if (!fp) {
        printf("Erro ao abrir o arquivo de relatório.\n");
        return;
    }

    // Relatório de Pacientes
    fprintf(fp, "Relatório de Pacientes:\n");
    Paciente *paciente = listaPacientes;
    while (paciente) {
        fprintf(fp, "ID: %d, Nome: %s, Idade: %d, Endereço: %s, Telefone: %s\n", paciente->id, paciente->nome, paciente->idade, paciente->endereco, paciente->telefone);
        paciente = paciente->prox;
    }

    // Relatório de Médicos
    fprintf(fp, "\nRelatório de Médicos:\n");
    Medico *medico = listaMedicos;
    while (medico) {
        fprintf(fp, "ID: %d, Nome: %s, Especialidade: %s\n", medico->id, medico->nome, medico->especialidade);
        medico = medico->prox;
    }

    // Relatório de Consultas
    fprintf(fp, "\nRelatório de Consultas:\n");
    Consulta *consulta = fila->inicio;
    while (consulta) {
        fprintf(fp, "Paciente ID: %d, Médico ID: %d, Gravidade: %d, Data: %s\n", consulta->idPaciente, consulta->idMedico, consulta->gravidade, consulta->data);
        consulta = consulta->prox;
    }

    fclose(fp);
    printf("Relatório gerado com sucesso em 'relatorio.txt'.\n");
}

void salvarDados(Paciente *pacientes, Medico *medicos, FilaPrioridade *fila) {
    FILE *fp;

    fp = fopen("pacientes.txt", "w");
    while (pacientes) {
        fprintf(fp, "%d %s %d %s %s\n", pacientes->id, pacientes->nome, pacientes->idade, pacientes->endereco, pacientes->telefone);
        pacientes = pacientes->prox;
    }
    fclose(fp);

    fp = fopen("medicos.txt", "w");
    while (medicos) {
        fprintf(fp, "%d %s %s\n", medicos->id, medicos->nome, medicos->especialidade);
        medicos = medicos->prox;
    }
    fclose(fp);
}

void carregarDados(Paciente **pacientes, Medico **medicos, FilaPrioridade *fila) {
    FILE *fp;
    int id, idade;
    char nome[MAX_NAME], endereco[MAX_NAME], telefone[MAX_NAME], especialidade[MAX_NAME];

    fp = fopen("pacientes.txt", "r");
    if (fp) {
        while (fscanf(fp, "%d %[^\n] %d %[^\n] %[^\n]\n", &id, nome, &idade, endereco, telefone) != EOF) {
            cadastrarPaciente(pacientes, id, nome, idade, endereco, telefone);
        }
        fclose(fp);
    }

    fp = fopen("medicos.txt", "r");
    if (fp) {
        while (fscanf(fp, "%d %[^\n] %[^\n]\n", &id, nome, especialidade) != EOF) {
            cadastrarMedico(medicos, id, nome, especialidade);
        }
        fclose(fp);
    }
}
